framework/packages/abstract-sdk/src/apis/authz.rs

```diff
@@ -0,0 +1,246 @@
+        let granter = granter.unwrap_or(self.proxy_address(deps)?);
+        Ok(AuthZ { granter })
+    }
+}
+
+impl<T> AuthZInterface for T where T: AccountIdentification {}
+
+/// This struct provides methods to grant message authorizations and interact with the authz module.
+///
+/// # Example
+/// ```
+/// use abstract_sdk::prelude::*;
+/// # use cosmwasm_std::testing::mock_dependencies;
+/// # use abstract_sdk::mock_module::MockModule;
+/// # let module = MockModule::new();
+///
+/// let authz: Authz  = module.auth_z(deps.as_ref(), None)?;
+/// ```
+/// */
+#[derive(Clone)]
+pub struct AuthZ {
+    granter: Addr,
+}
+
+impl AuthZ {
+    /// Retrieve the granter's address.
+    /// By default, this is the proxy address of the Account.
+    fn granter(&self) -> Addr {
+        self.granter.clone()
+    }
+
+    /// Removes msg type authorization from the granter to the **grantee**.
+    ///
+    /// # Arguments
+    ///
+    /// * `grantee` - The address of the grantee.
+    /// * `type_url` - The msg type url to revoke authorization.
+    pub fn revoke(&self, grantee: &Addr, type_url: String) -> CosmosMsg {
+        let msg = authz::v1beta1::MsgRevoke {
+            granter: self.granter().to_string(),
+            grantee: grantee.to_string(),
+            msg_type_url: type_url,
+        }
+        .encode_to_vec();
+
+        CosmosMsg::Stargate {
+            // TODO: `Name` implementation is missing for authz
+            // type_url: authz::v1beta1::MsgRevoke::type_url(),
+            type_url: "/cosmos.authz.v1beta1.MsgRevoke".to_string(),
+            value: Binary(msg),
+        }
+    }
+
+    /// Grants generic authorization to a **grantee**.
+    ///
+    /// # Arguments
+    ///
+    /// * `grantee` - The address of the grantee.
+    /// * `msg` - Allowed message type url.
+    pub fn grant_generic(
+        &self,
+        grantee: &Addr,
+        msg: String,
+        expiration: Option<Timestamp>,
+    ) -> CosmosMsg {
+        let generic = GenericAuthorization::new(msg);
+
+        let msg = authz::v1beta1::MsgGrant {
+            granter: self.granter().to_string(),
+            grantee: grantee.to_string(),
+            grant: Some(generic.grant(expiration)),
+        }
+        .encode_to_vec();
+
+        CosmosMsg::Stargate {
+            type_url: "/cosmos.authz.v1beta1.MsgGrant".to_string(),
+            value: Binary(msg),
+        }
+    }
+}
+
+fn convert_stamp(stamp: Timestamp) -> prost_types::Timestamp {
+    prost_types::Timestamp {
+        seconds: stamp.seconds() as i64,
+        nanos: stamp.nanos() as i32,
+    }
+}
+
+/// Represents a generic authorization grant.
+#[derive(Serialize, Deserialize, Clone, PartialEq, Eq, JsonSchema)]
+pub struct GenericAuthorization {
+    /// Allowed msg type_url
+    pub msg: String,
+}
+
+impl GenericAuthorization {
+    /// Create new generic authorization
+    pub fn new(msg: String) -> Self {
+        Self { msg }
+    }
+}
+
+trait AuthZAuthrization {
```

Comment:
What's the goal/point of this trait?

Reply:
Would be nice to have something that also accepts raw proto messages (`Vec<u8>`)

Reply:
If there are other grant types (besides GenericAuthorization) then we should also include those in here. And imo that should be in a nested mod

```rust
mod authorizations {
    // trait

    // different impls
}
```

Reply:
> What's the goal/point of this trait?

It helps structuring different types of grants

> If there are other grant types (besides GenericAuthorization) then we should also include those in here. And imo that should be in a nested mod
I will add other types of grants. Yeah, nested mod is fine